### PR TITLE
Add the support of the load RDB command

### PR DESCRIPTION
**Pull request by xq2010**
This PR is for https://github.com/apache/kvrocks/issues/1753

### CODE REVIEW

File: src/storage/rdb.h
Comment (git-hulk) [inline src/storage/rdb.h]: Can get config via storage's GetConfig.

**Comment by git-hulk** [comment]
Hi @xq2010 Thanks for your excellent contribution first!

About how to load the RDB file, I'm wondering if it's better to export a command to do this. For example, we can support a new command like `load $type $path` which allows loading RDB file from local.

```shell
$ redis-cli 6666> load RDB /path/to/your/rdb_file
```

So that we don't need to introduce a new tool. How do you feel about this? cc @apache/kvrocks-committers

**Comment by xq2010** [comment]
> Hi @xq2010 Thanks for your excellent contribution first!↳
> 
> About how to load the RDB file, I'm wondering if it's better to export a command to do this. For example, we can support a new command like `load $type $path` which allows loading RDB file from local.↳
> 
> ```shell
> $ redis-cli 6666> load RDB /path/to/your/rdb_file
> ```
> 
> So that we don't need to introduce a new tool. How do you feel about this? cc @apache/kvrocks-committers↳


 If we add the 'load rdb' command, we need to determine how to process an existing key that's present in both kvrocks and rdb, particularly when the key's type is a string in rdb. It will overwrite the existing key in kvrocks, which may be of a different type."

**Comment by git-hulk** [comment]
> If we add the 'load rdb' command, we need to determine how to process an existing key that's present in both kvrocks and rdb, particularly when the key's type is a string in rdb. It will overwrite the existing key in kvrocks, which may be of a different type."

Yes, it's a good concern. I prefer overwriting by default if it exists and we can offer an option NX like the set command to let users determine in the future. This is my personal perspective, would like to hear more voices about the loading way design.

**Comment by xq2010** [comment]
> Yes, it's a good concern. I prefer overwriting by default if it exists and we can offer an option NX like the set command to let users determine in the future. This is my personal perspective, would like to hear more voices about the loading way design.↳

Using NX is a good idea, I would also like to hear more opinions.

File: tests/testdata/corrupt_empty_keys.rdb
Comment (tisonkun) [inline tests/testdata/corrupt_empty_keys.rdb:1]: Generally, Apache release doesn't allow ship binary files. May you describe how these files are generated? Is it possible to generate test data file on the fly instead of committing them?

Comment (xq2010) [reply tests/testdata/corrupt_empty_keys.rdb:1]: These rdb files are copied from 'Redis's tests/asset/*.rdb'. If this is not allowed, I can generate the test data files on the fly

**Comment by PragmaTwice** [comment]
Thanks for your contribution!

Agree with @git-hulk, I think it would be better to just implement it like a command inside the kvrocks process instead of a new program.

**Comment by xq2010** [comment]
@git-hulk @PragmaTwice Okay, I will add a command to load the RDB file. The command might be rdb load \<path\> [namespace]. Do you have any suggestions or advice?

**Comment by git-hulk** [comment]
> @git-hulk @PragmaTwice Okay, I will add a command to load the RDB file. The command might be rdb load <path> [namespace]. Do you have any suggestions or advice?

Thank you! Sounds good to me, but [namespace] seems unnecessary since it knows the current namespace once it enters the command.

**Comment by xq2010** [comment]
> Thank you! Sounds good to me, but [namespace] seems unnecessary since it knows the current namespace once it enters the command.↳

Yes, it's unnecessary to use a namespace.

If the RDB file contains multiple db, the current implementation sets the namespace of each db to \<ns\>_\<db number\>. The password will be the same as the namespace. Only db 0 will use the current namespace. 
Perhaps using a parameter to specify which db to load would be better. If so, the command would be rdb load \<path\> [database number (default 0)].

**Comment by git-hulk** [comment]
> Perhaps using a parameter to specify which db to load would be better. If so, the command would be rdb load <path> [database number (default 0)].

@xq2010 Yes, I think this is a good solution.

File: tests/testdata/corrupt_empty_keys.rdb
Comment (tisonkun) [reply tests/testdata/corrupt_empty_keys.rdb:1]: Yes please.

**Comment by git-hulk** [comment]
@xq2010 Cool! The overall design is nice. I'll take another pass to take a look at detail implementation tomorrow.

File: CMakeLists.txt
Comment (tisonkun) [inline CMakeLists.txt]: Please revert this change.

**Review by tisonkun — COMMENTED** [review]
Generally looks good. But I don't have time to check in details for giving an approval on the implementation 🤣

File: src/storage/rdb.cc
Comment (git-hulk) [inline src/storage/rdb.cc:393]: Why do we need to add NOLINT for this line?

File: src/storage/rdb.cc
Comment (git-hulk) [inline src/storage/rdb.cc]:
```suggestion
  }

  return {Status::RedisParseErr, fmt::format("unsupported type: {}", type)};
}
```

File: src/storage/rdb.cc
Comment (git-hulk) [inline src/storage/rdb.cc]: Can we bring the error message into Status?

File: src/storage/rdb.cc
Comment (git-hulk) [inline src/storage/rdb.cc]: The same as the above

File: src/storage/rdb.h
Comment (git-hulk) [inline src/storage/rdb.h]: exist => isKeyExisted

File: src/storage/rdb.cc
Comment (git-hulk) [inline src/storage/rdb.cc]: I think we can print the empty keys info, no matter it's zero or not.

File: tests/cppunit/rdb_test.cc
Comment (git-hulk) [inline tests/cppunit/rdb_test.cc:312]:
```suggestion
}

```

File: tests/cppunit/rdb_util.h
Comment (git-hulk) [inline tests/cppunit/rdb_util.h:203]:
```suggestion
};

```

File: tests/cppunit/rdb_stream_test.cc
Comment (git-hulk) [inline tests/cppunit/rdb_stream_test.cc]:
```suggestion
}

```

File: src/storage/rdb.cc
Comment (git-hulk) [inline src/storage/rdb.cc]:
```suggestion
}

```

File: src/storage/rdb.cc
Comment (git-hulk) [inline src/storage/rdb.cc]: @PragmaTwice What do you think about this implementation.

File: src/commands/cmd_server.cc
Comment (torwig) [inline src/commands/cmd_server.cc]: I'm wondering if we can rename this parameter to `overwrite_existing` that will be `true` by default if I understood the meaning of the `is_nx` parameter.

File: src/common/rdb_stream.h
Comment (torwig) [inline src/common/rdb_stream.h]: I guess we can use `<cstdint>` here as a C++-header instead of C-like `<stdint.h>`.

File: src/storage/rdb.cc
Comment (PragmaTwice) [reply src/storage/rdb.cc]: I think it can just be replaced by something like `void logWhenError(v) { if (!v) LOG... }`.

We'd better to keep as less macros as we can.

File: src/commands/cmd_server.cc
Comment (PragmaTwice) [inline src/commands/cmd_server.cc]: Could you make a little explanation here about where this pointer need to be "share" for undetermined lifetime? Otherwise we'd better to use `unique_ptr` : )

File: src/commands/cmd_server.cc
Comment (PragmaTwice) [inline src/commands/cmd_server.cc]: Seems these variables are useless?

File: src/common/rdb_stream.cc
Comment (torwig) [inline src/common/rdb_stream.cc]: Perhaps, we can shorten this line by using `std::min(max_read_chunk_size_, len)`.

File: src/storage/rdb.cc
Comment (torwig) [inline src/storage/rdb.cc]: I think we can make the value `11` a constant.

File: src/storage/rdb.cc
Comment (torwig) [inline src/storage/rdb.cc]: Perhaps here we also can make `2` and `8` constants (to describe that something is a checksum padding and something is a checksum length).

File: src/storage/rdb.cc
Comment (torwig) [inline src/storage/rdb.cc]: The same is related to `4` here and `8`  down the code.

File: src/storage/rdb.cc
Comment (torwig) [inline src/storage/rdb.cc]: You can name the function more descriptive like `loadExpirationTimeSeconds`.

File: src/storage/rdb.cc
Comment (torwig) [inline src/storage/rdb.cc]: You can name the function more descriptive like `loadExpirationTimeMilliseconds`.

File: src/storage/rdb.cc
Comment (torwig) [inline src/storage/rdb.cc:563]:
```
uint64_t expire_time = 0;
int64_t expire_keys = 0;
int64_t load_keys = 0;
int64_t empty_keys_skipped = 0;
uint64_t skip_exist_keys = 0;
```

Minor renaming to the following:
```
uint64_t expiration_time = 0;
int64_t expired_keys = 0;
int64_t loaded_keys = 0;
int64_t skipped_empty_keys = 0;
uint64_t skipped_existing_keys = 0;
```

File: src/storage/rdb.cc
Comment (torwig) [inline src/storage/rdb.cc]: Make `5` as a constant (something like `kMinRdbVersionToVerifyChecksum` or something shorter).

File: src/storage/rdb.cc
Comment (torwig) [inline src/storage/rdb.cc]: Transform `1` into the `kMinRdbVersion` constant.

File: src/storage/rdb.cc
Comment (torwig) [inline src/storage/rdb.cc]: So we check the checksum to return an error? What would be the result? As I understand from the code, all objects will be loaded and the function `RDB::LoadRdb` will return an error without any rollback. Can we write something like "All objects were processed and loaded but the checksum is unexpected" in the response to the client?

File: src/storage/rdb.cc
Comment (torwig) [inline src/storage/rdb.cc]: I would rename this function to `keyExist` or as @git-hulk suggested `isKeyExist`.

File: src/storage/rdb.cc
Comment (torwig) [inline src/storage/rdb.cc]: I was thinking of introducing the function `Database::KeyExist` to make the check, in this case, simpler without creating `std::vector` and renaming `Database::Exists` to `Database::MultipleKeysExist`. However, I'm not sure about it. @git-hulk @PragmaTwice What do you think?

**Review by torwig — COMMENTED** [review]
@xq2010 Good job, thank you for your effort. Just a little nitpicking from me :)

File: src/common/rdb_stream.h
Comment (PragmaTwice) [inline src/common/rdb_stream.h:39]: not sure if we need to consider endian here.

File: src/storage/rdb.cc
Comment (git-hulk) [reply src/storage/rdb.cc]: Yes,this solution will be better.

**Comment by xq2010** [comment]
Thanks for your review, I will review it later

File: src/storage/rdb.h
Comment (xq2010) [reply src/storage/rdb.h]: The last commit has deleted the config parameter

File: src/storage/rdb.cc
Comment (xq2010) [reply src/storage/rdb.cc:393]: The lint report falsely indicates a memory leak, but I can't find any potential for a memory leak

File: src/storage/rdb.cc
Comment (xq2010) [reply src/storage/rdb.cc]: yes, accepted

File: src/storage/rdb.cc
Comment (xq2010) [reply src/storage/rdb.cc]: okay

File: src/storage/rdb.cc
Comment (xq2010) [reply src/storage/rdb.cc]: Okay

File: src/commands/cmd_server.cc
Comment (xq2010) [reply src/commands/cmd_server.cc]: Yes, it's better to use unique_ptr.

File: src/commands/cmd_server.cc
Comment (xq2010) [reply src/commands/cmd_server.cc]: Yes, I copied it and forgot to delete.

File: src/common/rdb_stream.cc
Comment (xq2010) [reply src/common/rdb_stream.cc]: Okay

File: src/common/rdb_stream.h
Comment (xq2010) [reply src/common/rdb_stream.h:39]: I don't think so. It only reads one byte. If you plan to use it to read multiple bytes, you should determine how to handle endianness

File: src/storage/rdb.cc
Comment (git-hulk) [reply src/storage/rdb.cc:393]: ok, it should be if the ASAN in CI is good.

File: src/common/rdb_stream.cc
Comment (mapleFU) [inline src/common/rdb_stream.cc:40]: It's a bit tricky here. Should we `memrev64ifbe` for the crc64 of the Stream?

File: src/common/rdb_stream.cc
Comment (mapleFU) [inline src/common/rdb_stream.cc]: using `static_cast` for  `buf`?

File: src/common/rdb_stream.cc
Comment (mapleFU) [inline src/common/rdb_stream.cc:58]: https://en.cppreference.com/w/cpp/io/basic_istream/read
Do we need to handle the exception here?

File: src/common/rdb_stream.h
Comment (mapleFU) [inline src/common/rdb_stream.h]: Actually I think `GetCheckSum()` here is a bit tricky. Can we use `std::optional<uint64_t> check_sum_` as member, and return `Error` when `check_sum_` is unset?

File: src/common/rdb_stream.h
Comment (PragmaTwice) [reply src/common/rdb_stream.h:39]: Oh, didn't notice it. Looks fine then.

File: src/storage/rdb.cc
Comment (xq2010) [reply src/storage/rdb.cc:563]: I believe the original version is better :)

File: src/storage/rdb.cc
Comment (xq2010) [reply src/storage/rdb.cc]: Yes, if the checksum does not match, we will not perform a rollback. In Redis, if a checksum error occurs, the server will exit during start-up. However, as a command to load the RDB, it is not suitable to exit. Instead, I will add an error message to be sent to the client.

File: src/common/rdb_stream.cc
Comment (xq2010) [reply src/common/rdb_stream.cc:40]: yes, we should, , that's what Redis does.

File: src/common/rdb_stream.h
Comment (xq2010) [reply src/common/rdb_stream.h]: i don't think so, if not set, it should not return an error

**Comment by git-hulk** [comment]
Hi @xq2010 How's this PR going? Should we take another round review?

**Comment by xq2010** [comment]
@git-hulk  Thank you for your review. I've made some corrections. Could you please review it once more?

**Comment by git-hulk** [comment]
> @git-hulk Thank you for your review. I've made some corrections. Could you please review it once more?

Sure, thanks for your efforts!

**Review by git-hulk — APPROVED** [review]

**Comment by git-hulk** [comment]
I would like to merge this PR today if no more comments.

**Comment by mapleFU** [comment]
I'm a bit busy these few days. I'll go through it before Thursday. You can merge this first if other doesn't have any comments.

**Review by torwig — APPROVED** [review]
Generally, LGTM.

File: src/common/rdb_stream.cc
Comment (mapleFU) [inline src/common/rdb_stream.cc:26]: The syntax of `RdbStringStream` is different from `RdbFileStream`. The previous one will return a smaller size when n is greater than remaining buffer, but `RdbFileStream` will always return `Status::NotOK`. Is this expected?

File: src/storage/rdb.cc
Comment (mapleFU) [inline src/storage/rdb.cc:177]: Could it be `len == 0`?

File: src/common/rdb_stream.cc
Comment (mapleFU) [reply src/common/rdb_stream.cc:26]: Oh I make it wrong. So they're same.

File: src/storage/rdb.cc
Comment (xq2010) [reply src/storage/rdb.cc:177]: Yes, it is possible, could you  help me fix it?

Comment (mapleFU) [reply src/storage/rdb.cc:177]: https://github.com/apache/kvrocks/pull/1839
I've create a pull request for it.

Thanks again for your contribution!

**Pull request merged**
